library/core/src/num/dec2flt/decimal.rs

```diff
@@ -0,0 +1,351 @@
+//! Arbitrary-precision decimal class for fallback algorithms.
+//!
+//! This is only used if the fast-path (native floats) and
+//! the Eisel-Lemire algorithm are unable to unambiguously
+//! determine the float.
+//!
+//! The technique used is "Simple Decimal Conversion", developed
+//! by Nigel Tao and Ken Thompson. A detailed description of the
+//! algorithm can be found in "ParseNumberF64 by Simple Decimal Conversion",
+//! available online: <https://nigeltao.github.io/blog/2020/parse-number-f64-simple.html>.
+
+use crate::num::dec2flt::common::{is_8digits, parse_digits, ByteSlice, ByteSliceMut};
+
+#[derive(Clone)]
+pub struct Decimal {
+    /// The number of significant digits in the decimal.
+    pub num_digits: usize,
+    /// The offset of the decimal point in the significant digits.
+    pub decimal_point: i32,
+    /// If the number of significant digits stored in the decimal is truncated.
+    pub truncated: bool,
```

Comment:
It would be nice if we used this field to emit a warning when the parsed value bitwise diverges from what the user wrote. Not necessary for this PR.

Reply:
I'll submit another PR then later, since this one is relatively gigantic. That being said, it might be a bit more complex to do, since the question is "what differs". `let f: f64 = 0.1;` differs from what the user wrote, but it's very close to the nearest possible float. Also, there is no exact representation: do we use `0.10000000000000000555`, the rounding to 20 decimal places? If someone writes `let f: f64 = 0.10000000000000000555`, since by default it formats to `0.1`, do we warn of the difference, even though the representation is quite exact?

It might be doable by limiting the output to the number of input digits, with the same exponent as the input, but that would likely have to be handled elsewhere. 

I'm assuming the best way would be as follows, which is mostly handled by the tokenizer:

1. Get if there's exponent notation.
2. Count the number of significant digits.
3. If there's scientific notation, check if the output is identical to `format!("{:.*e}", digits - 1, f);`, after removing the `+` sign digits, and shifting digits from the exponent to the significant digits or vice versa as necessary.
4. If there isn't, check if the output is identical to `format!("{:.*}", digits, f);`.

Note, that we do need to shift digits, since cases like:

```rust
let f = 12345e10;
```

Will be formatted as:

```rust
"1.2345e14"
```

Which obviously, would require us to shift the decimal place to ensure the same exponent (shift the decimal point 4-places right, and since the decimal point is missing in the input, remove it). Currently, the tokenizer produces:

```rust
/// "12.34f32", "0b100.100"
Float { base: Base, empty_exponent: bool },
```

So we would need likely to have more information here.